### PR TITLE
ShaderGen: Get rid of static buffers

### PR DESCRIPTION
**Pull request by lioncash**

### CODE REVIEW

File: Source/Core/VideoCommon/ShaderGenCommon.h [hunk 1]
```diff
@@ -62,6 +59,9 @@ class ShaderGeneratorInterface
 	 */
 	template<class uid_data>
 	uid_data* GetUidData() { return nullptr; }
+
+protected:
+	std::string m_buffer;
```
Comment (degasus) [inline Source/Core/VideoCommon/ShaderGenCommon.h]: Please reserve a bigger buffer within the constructor. Operator+= isn't the best way to grow this buffer often.

Comment (lioncash) [inline Source/Core/VideoCommon/ShaderGenCommon.h]: The reservation is done in ShaderCode, as ShaderUid doesn't use writing, so it doesn't make sense to pre-allocate for that.

Comment (degasus) [inline Source/Core/VideoCommon/ShaderGenCommon.h]: Oh, this is within the superclass? Are you sure std::string() has no overhead? Sounds fine to me, but maybe check if the assembly of the UID generator is still the same.

Comment (lioncash) [inline Source/Core/VideoCommon/ShaderGenCommon.h]: I'm quite sure an empty string won't have adverse effects. Anyhow, it's kind of odd to utilize a class intended for generating/writing shader code as the base for UID generation as well.

**Comment by degasus** [comment]
LGTM

**Comment by lioncash** [comment]
@dolphin-emu-bot rebuild

**Comment by phire** [comment]
I don't really have time to debug, but I suspect this backtrace from spxtr belongs here:

Was fixed by deleting the shader cache.

    (gdb) backtrace
    #0  0x00007fffefac3aa4 in free () from /usr/lib/libc.so.6
    #1  0x000000000077b187 in __gnu_cxx::new_allocator<char>::deallocate (this=<optimized out>, __p=<optimized out>) at /usr/include/c++/5.3.0/ext/new_allocator.h:110
    #2  std::allocator_traits<std::allocator<char> >::deallocate (__a=..., __n=<optimized out>, __p=<optimized out>) at /usr/include/c++/5.3.0/bits/alloc_traits.h:386
    #3  std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_destroy (__size=<optimized out>, this=0x7fffd7313aa0) at /usr/include/c++/5.3.0/bits/basic_string.h:185
    #4  std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_dispose (this=0x7fffd7313aa0) at /usr/include/c++/5.3.0/bits/basic_string.h:180
    #5  std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string (this=0x7fffd7313aa0, __in_chrg=<optimized out>) at /usr/include/c++/5.3.0/bits/basic_string.h:544
    #6  ShaderGeneratorInterface::~ShaderGeneratorInterface (this=0x7fffd7313a98, __in_chrg=<optimized out>) at /home/dash/dolphin/Source/Core/VideoCommon/ShaderGenCommon.h:31
    #7  ShaderUid<geometry_shader_uid_data>::~ShaderUid (this=0x7fffd7313a98, __in_chrg=<optimized out>) at /home/dash/dolphin/Source/Core/VideoCommon/ShaderGenCommon.h:74
    #8  OGL::SHADERUID::~SHADERUID (this=0x7fffd7313900, __in_chrg=<optimized out>) at /home/dash/dolphin/Source/Core/VideoBackends/OGL/ProgramShaderCache.h:19
    #9  LinearDiskCache<OGL::SHADERUID, unsigned char>::OpenAndRead (reader=..., filename="/home/dash/.dolphin-emu/Cache/Shaders/ogl-GALE01-shaders.cache", this=0x1457c80 <OGL::g_program_disk_cache>)
    at /home/dash/dolphin/Source/Core/Common/LinearDiskCache.h:72
    #10 OGL::ProgramShaderCache::Init () at /home/dash/dolphin/Source/Core/VideoBackends/OGL/ProgramShaderCache.cpp:447
    #11 0x00000000007768da in OGL::VideoBackend::Video_Prepare (this=<optimized out>) at /home/dash/dolphin/Source/Core/VideoBackends/OGL/main.cpp:195
    #12 0x00000000005b2409 in Core::EmuThread () at /home/dash/dolphin/Source/Core/Core/Core.cpp:533
    #13 0x00007fffedfecf40 in std::(anonymous namespace)::execute_native_thread_routine (__p=<optimized out>) at /build/gcc/src/gcc-5.3.0/libstdc++-v3/src/c++11/thread.cc:84
    #14 0x00007fffeec844a4 in start_thread () from /usr/lib/libpthread.so.0
    #15 0x00007fffefb3113d in clone () from /usr/lib/libc.so.6

**Comment by degasus** [comment]
@phire Sounds like we memset our std::string somewhere...

**Comment by lioncash** [comment]
If anything, the interface itself is an issue, because something related to shader generation is directly tacking UID generation into it, rather than being a composed part of it.

File: Source/Core/VideoBackends/OGL/FramebufferManager.cpp [hunk 2]
```diff
@@ -347,8 +347,8 @@ FramebufferManager::FramebufferManager(int targetWidth, int targetHeight, int ms
 		"	}\n"
 		"}\n";
 
-	ProgramShaderCache::CompileShader(m_pixel_format_shaders[0], vs, ps_rgb8_to_rgba6.c_str(), (m_EFBLayers > 1) ? gs.c_str() : nullptr);
-	ProgramShaderCache::CompileShader(m_pixel_format_shaders[1], vs, ps_rgba6_to_rgb8.c_str(), (m_EFBLayers > 1) ? gs.c_str() : nullptr);
+	ProgramShaderCache::CompileShader(m_pixel_format_shaders[0], vs, ps_rgb8_to_rgba6.c_str(), (m_EFBLayers > 1) ? gs : "");
+	ProgramShaderCache::CompileShader(m_pixel_format_shaders[1], vs, ps_rgba6_to_rgb8.c_str(), (m_EFBLayers > 1) ? gs : "");
```
Comment (imirkin) [inline Source/Core/VideoBackends/OGL/FramebufferManager.cpp]: Should the above ps_* strings have lost their c_str()'s?

File: Source/Core/VideoBackends/OGL/TextureCache.cpp [hunk 3]
```diff
@@ -413,14 +413,14 @@ void TextureCache::CompileShaders()
...
 
-	const char* prefix = (GProgram == nullptr) ? "f" : "v";
-	const char* depth_layer = (g_ActiveConfig.bStereoEFBMonoDepth) ? "0.0" : "f_uv0.z";
+	const char* prefix = geo_program.empty() ? "f" : "v";
+	const char* depth_layer = g_ActiveConfig.bStereoEFBMonoDepth ? "0.0" : "f_uv0.z";
 
-	ProgramShaderCache::CompileShader(s_ColorCopyProgram, StringFromFormat(VProgram, prefix, prefix).c_str(), pColorCopyProg, GProgram);
-	ProgramShaderCache::CompileShader(s_ColorMatrixProgram, StringFromFormat(VProgram, prefix, prefix).c_str(), pColorMatrixProg, GProgram);
-	ProgramShaderCache::CompileShader(s_DepthMatrixProgram, StringFromFormat(VProgram, prefix, prefix).c_str(), StringFromFormat(pDepthMatrixProg, depth_layer).c_str(), GProgram);
+	ProgramShaderCache::CompileShader(s_ColorCopyProgram, StringFromFormat(vertex_program, prefix, prefix).c_str(), color_copy_program, geo_program);
```
Comment (imirkin) [inline Source/Core/VideoBackends/OGL/TextureCache.cpp]: And remove the c_str's on these too?

**Pull request merged**